Upgrade to .Net Core v3.0.0-preview2

Brar opened the pull request:
As mentioned in #806 this would be necessary to create some basic compatibility with .Net Core v3.0.0-preview2 which has been released 2019-01-29

I recognize there are quite a few (66 on my machine) test failures that I can't really trace down to my changes but that might be interesting.

I'm marking this as draft PR as in its current state it certainly isn't ready to merge but I'm interested in feedback anyways.

roji commented on src/EFCore.PG/Infrastructure/Internal/NpgsqlOptionsExtension.cs:
```diff
@@ -195,6 +195,12 @@ public virtual NpgsqlOptionsExtension WithRemoteCertificateValidationCallback([C
 
+        public override void PopulateDebugInfo(IDictionary<string, string> debugInfo)
+        {
+            // We migth want to add the current values of AdminDatabase, PostgresVersion, ReverseNullOrdering, _userRangeDefinitions
```
Yeah, in theory I think we need to include everything that's configurable on the options extension.

nit: typo (`migth`)

Brar replied:
Ok.
I think that implementing this will be my next commit.
Question: Should we populate debugInfo only if the values have been set to something different than their default value or should we put in default values too?

roji replied:
I'm not sure, but we should follow whatever the EF Core providers do. I seem to remember that they include information even if it's the default - we should confirm that.

Brar replied:
The [SQL Server provider](https://github.com/aspnet/EntityFrameworkCore/blob/025a5589b25d4b96318d4ba2d4e9aed0d32e527d/src/EFCore.SqlServer/Infrastructure/Internal/SqlServerOptionsExtension.cs#L88) and the [Sqlite provider](https://github.com/aspnet/EntityFrameworkCore/blob/ca6f9a8ef8aa4bea9e8cb085cf5e5bd21cf84ba7/src/EFCore.Sqlite.Core/Infrastructure/Internal/SqliteOptionsExtension.cs#L85) both log unconditionally. On the other hand they have few to no configurable options to log anyways.

roji replied:
Good point. I have no idea what the effect would be of always adding all the options, but I'm guessing it could make logging much heavier. It may make sense to only add info when it deviates from the default.

roji commented on src/EFCore.PG.NodaTime/Infrastructure/Internal/NpgsqlNodaTimeOptionsExtension.cs:
```diff
@@ -23,6 +23,10 @@ public virtual bool ApplyServices(IServiceCollection services)
 
         public virtual long GetServiceProviderHashCode() => 0;
 
+        public void PopulateDebugInfo(IDictionary<string, string> debugInfo)
```
Seems like we should follow SQL Server/Sqlite here and populate like them (see their own corresponding files for an example).

Brar commented:
> Do you want to to work on this and have time? If not I'll have to get into anyway soon...

I can say that I'm interested in it (I'm experimenting with .Net Core 3.0 anyways) and that I have at least some time that I can invest here.

I'll keep pushing necessary changes to my repository so that they'll appear here.

In the best case I can get this to a state where we can merge it.
In the worst case you can possibly cherry-pick some of the changes when you are starting your work to upgrade to preview3.

roji commented:
OK, thanks for the help @Brar! If at all possible, try to separate commits as I wrote above, this will make it easier to understand what's going on and to cherry-pick.

Brar commented:
Ok @roji, I've split up the previous commit into several smaller ones (I obviously needed to force-push that) and tried to document the reasons behind the changes in the commit messages.
We can always squash them or some of them if it gets too much.

There is no real change in the code with regard to the version you've already reviewed (except that I fixed the typo).

Is that the way that would be helpful if you need to pick those changes or is there something I can improve?

roji commented:
@Brar thanks. I'll probably start working on syncing to preview3 myself in the coming days, since the release is drawing near. FYI after preview3 I intend to try to keep the Npgsql provider in sync much more regularly, rather than a big-bang sync session just before release. That should make the work easier.

roji commented:
@Brar, probably starting tomorrow I'll work on a fully compatible preview3 release - I'll start by looking and taking your commits and I'll let you know how it goes! It's definitely great to have some help to start on this.

Brar commented:
@roji, great!
I'll push a commit with a `PopulateDebugInfo(IDictionary<string, string> debugInfo)` implementation soon (tomorrow?).

I've figured out, that the content of the `debugInfo` dictionary isn't supposed to be human readable but is meant to enable automated diffs. Other providers set the key to an unique string, representing the configurable option and the value to the stringified hash code of its current value.
See [CoreOptionsExtensions](https://github.com/aspnet/EntityFrameworkCore/blob/590e5f8e214b9ae1028615ba81f9b0cfc39ac964/src/EFCore/Infrastructure/CoreOptionsExtension.cs#L392) for a pretty complete example.

I didn't find the tests for this code yet, but I want to make sure that there really aren't any tests covering this before moving forward.

I also found out that we should probably override `RelationalOptionsExtension.LogFragment` which seems to be meant for human readable logging of configuable options.
I've uncommented `LoggingNpgsqlTest.Logs_context_initialization_admin_database()` and got it to pass after overriding `RelationalOptionsExtension.LogFragment`.
But there is more work to do for the other options beyond `AdminDatabase`.
I'll look into this too, if you want me to.

roji commented:
@Brar sure, I'll leave all the debugging/logging part to you if you want, and work on making everything else build and pass for preview3.

Brar commented:
Ok, strictly speaking the latest commit (`Override LogFragment in NpgsqlOptionsExtension`) is out of context for this PR as it's got nothing to do with upgrading to .Net Core v3.0.0-preview2 but it was a low hanging fruit which I've stumbled upon while investigating how to test `PopulateDebugInfo` so I just went ahead and implemented it.

The `PopulateDebugInfo` implementation will follow later this evening.

Brar commented:
Regarding `PopulateDebugInfo` I've verified that there is no test in the framework that specifically tests its functionality (maybe there should be?), although its code is covered by lots of tests (none of which fails after I've added the method body).

@roji I think that will be all for today.
You can pick whatever you deem useful.
I'll also happily change/fix things upon another review.

roji commented:
Thanks @Brar, I already started a bit of work in https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/tree/preview3 (I already did some fixes to make more tests pass), but I'll probably wait a few more days until the Microsoft preview3 release happens. In any case I'll look at your work and cherry-pick it too, thanks!

KexyBiscuit commented:
[.NET Core 3.0 Preview3 19153.1](https://github.com/aspnet/EntityFrameworkCore/releases/tag/3.0.0-preview3.19153.1) is out now, when will this get merged?

roji commented:
@KexyBiscuit you can expect a preview3 version to be out in the next day or two.

roji commented:
Superceded by #844
